File: awiki/nodejs-client/tests/cross_platform/test_plain_messages.js
#!/usr/bin/env node

/**
 * Cross-platform plain text message test.
 * 
 * Tests:
 * 1. Python -> Node.js plain message
 * 2. Node.js -> Python plain message
 * 
 * Usage:
 *   node tests/cross_platform/test_plain_messages.js --from python1 --to node1
 *   node tests/cross_platform/test_plain_messages.js --from node1 --to python1
 */

import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import axios from 'axios';
import { createSDKConfig } from '../../src/utils/config.js';
import { loadIdentity } from '../../src/credential_store.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PYTHON_CRED_DIR = join(process.env.USERPROFILE || '', '.openclaw', 'credentials', 'awiki-agent-id-message');
const NODEJS_CRED_DIR = join(__dirname, '..', '.credentials');
const ALT_NODEJS_CRED_DIR = join(__dirname, '..', '..', '..', 'nodejs-awiki', '.credentials');

const TEST_RESULTS = {
    passed: 0,
    failed: 0,
    details: []
};

/**
 * Load credential from appropriate directory.
 */
function loadCredential(name) {
    // Try Node.js directory first
    try {
        const nodePath = join(NODEJS_CRED_DIR, `${name}.json`);
        return JSON.parse(readFileSync(nodePath, 'utf-8'));
    } catch (e) {
        // Try alternative Node.js directory
        try {
            const altNodePath = join(ALT_NODEJS_CRED_DIR, `${name}.json`);
            return JSON.parse(readFileSync(altNodePath, 'utf-8'));
        } catch (e2) {
            // Try Python directory
            try {
                const pythonPath = join(PYTHON_CRED_DIR, `${name}.json`);
                return JSON.parse(readFileSync(pythonPath, 'utf-8'));
            } catch (e3) {
                throw new Error(`Credential '${name}' not found in either directory`);
            }
        }
    }
}

/**
 * Send message using JWT.
 */
async function sendMessage(senderCred, receiverDid, content) {
    const config = createSDKConfig();
    
    const requestBody = {
        jsonrpc: '2.0',
        method: 'send',
        params: {
            sender_did: senderCred.did,
            receiver_did: receiverDid,
            content: content,
            type: 'text',
            client_msg_id: `test_${Date.now()}`
        },
        id: 1
    };
    
    try {
        const response = await axios.post(
            `${config.molt_message_url}/message/rpc`,
            requestBody,
            {
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${senderCred.jwt_token}`
                },
                timeout: 30000
            }
        );
        
        const result = response.data;
        
        if (result.result) {
            return {
                success: true,
                server_seq: result.result.server_seq,
                message_id: result.result.id
            };
        } else {
            return {
                success: false,
                error: result.error?.message || 'Unknown error'
            };
        }
    } catch (e) {
        return {
            success: false,
            error: e.message
        };
    }
}

/**
 * Check inbox for messages.
 */
async function checkInbox(receiverCred, expectedSender) {
    const config = createSDKConfig();
    
    const requestBody = {
        jsonrpc: '2.0',
        method: 'get_inbox',
        params: {
            user_did: receiverCred.did,
            limit: 10
        },
        id: 1
    };
    
    try {
        const response = await axios.post(
            `${config.molt_message_url}/message/rpc`,
            requestBody,
            {
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${receiverCred.jwt_token}`
                },
                timeout: 30000
            }
        );
        
        const result = response.data;
        
        if (result.result && result.result.messages) {
            // Find message from expected sender
            const expectedMsg = result.result.messages.find(
                msg => msg.sender_did === expectedSender
            );
            
            return {
                success: true,
                message: expectedMsg,
                total_messages: result.result.messages.length
            };
        } else {
            return {
                success: false,
                error: result.error?.message || 'No messages'
            };
        }
    } catch (e) {
        return {
            success: false,
            error: e.message
        };
    }
}

/**
 * Run cross-platform plain message test.
 */
async function runTest(fromName, toName) {
    console.log('='.repeat(60));
    console.log(`Cross-Platform Plain Message Test: ${fromName} -> ${toName}`);
    console.log('='.repeat(60));
    
    // Load credentials
    console.log('\n[1/4] Loading credentials...');
    let senderCred, receiverCred;
    
    try {
        senderCred = loadCredential(fromName);
        console.log(`  ✓ Sender (${fromName}): ${senderCred.did}`);
    } catch (e) {
        console.log(`  ✗ Failed to load sender credential: ${e.message}`);
        TEST_RESULTS.failed++;
        TEST_RESULTS.details.push({
            test: `Load sender credential (${fromName})`,
            success: false,
            error: e.message
        });
        return;
    }
    
    try {
        receiverCred = loadCredential(toName);
        console.log(`  ✓ Receiver (${toName}): ${receiverCred.did}`);
    } catch (e) {
        console.log(`  ✗ Failed to load receiver credential: ${e.message}`);
        TEST_RESULTS.failed++;
        TEST_RESULTS.details.push({
            test: `Load receiver credential (${toName})`,
            success: false,
            error: e.message
        });
        return;
    }
    
    // Send message
    console.log('\n[2/4] Sending message...');
    const testMessage = `[CROSS-PLATFORM TEST] ${fromName} -> ${toName} @ ${new Date().toISOString()}`;
    
    const sendResult = await sendMessage(senderCred, receiverCred.did, testMessage);
    
    if (sendResult.success) {
        console.log(`  ✓ Message sent (Server Seq: ${sendResult.server_seq})`);
        TEST_RESULTS.passed++;
        TEST_RESULTS.details.push({
            test: 'Send message',
            success: true,
            server_seq: sendResult.server_seq
        });
    } else {
        console.log(`  ✗ Failed to send: ${sendResult.error}`);
        TEST_RESULTS.failed++;
        TEST_RESULTS.details.push({
            test: 'Send message',
            success: false,
            error: sendResult.error
        });
        return;
    }
    
    // Wait for message delivery
    console.log('\n[3/4] Waiting for message delivery...');
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    // Check inbox
    console.log('\n[4/4] Checking receiver inbox...');
    const inboxResult = await checkInbox(receiverCred, senderCred.did);
    
    if (inboxResult.success && inboxResult.message) {
        const msg = inboxResult.message;
        
        // Verify message content
        if (msg.content.includes('[CROSS-PLATFORM TEST]')) {
            console.log(`  ✓ Message received and verified`);
            console.log(`    From: ${msg.sender_did}`);
            console.log(`    Content: ${msg.content.substring(0, 50)}...`);
            console.log(`    Server Seq: ${msg.server_seq}`);
            
            TEST_RESULTS.passed++;
            TEST_RESULTS.details.push({
                test: 'Receive and verify message',
                success: true,
                server_seq: msg.server_seq,
                content_match: true
            });
        } else {
            console.log(`  ✗ Message content mismatch`);
            TEST_RESULTS.failed++;
            TEST_RESULTS.details.push({
                test: 'Receive and verify message',
                success: false,
                error: 'Content mismatch'
            });
        }
    } else {
        console.log(`  ✗ Failed to receive message: ${inboxResult.error}`);
        TEST_RESULTS.failed++;
        TEST_RESULTS.details.push({
            test: 'Receive and verify message',
            success: false,
            error: inboxResult.error
        });
    }
}

/**
 * Parse command line arguments.
 */
function parseArgs() {
    const args = process.argv.slice(2);
    const result = {
        from: null,
        to: null
    };
    
    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--from':
                result.from = args[++i];
                break;
            case '--to':
                result.to = args[++i];
                break;
            case '--help':
            case '-h':
                console.log(`
Cross-Platform Plain Message Test

Usage:
  node tests/cross_platform/test_plain_messages.js --from <name> --to <name>

Examples:
  node tests/cross_platform/test_plain_messages.js --from python1 --to node1
  node tests/cross_platform/test_plain_messages.js --from node1 --to python1
`);
                process.exit(0);
                break;
        }
    }
    
    return result;
}

// Main
const options = parseArgs();

if (!options.from || !options.to) {
    console.error('Error: --from and --to are required');
    console.error('Use --help for usage information');
    process.exit(1);
}

await runTest(options.from, options.to);

// Print summary
console.log('\n' + '='.repeat(60));
console.log('TEST SUMMARY');
console.log('='.repeat(60));
console.log(`Passed: ${TEST_RESULTS.passed}`);
console.log(`Failed: ${TEST_RESULTS.failed}`);
console.log('='.repeat(60));

if (TEST_RESULTS.failed === 0) {
    console.log('\n✓ TEST PASSED');
    process.exit(0);
} else {
    console.log('\n✗ TEST FAILED');
    process.exit(1);
}

File: awiki/python-client.backup.20260308/scripts/test_plain_messages.py
#!/usr/bin/env python3
"""
Test plain text message sending between identities.
"""

import json
import asyncio
from pathlib import Path
import httpx

print("=" * 80)
print("Python: Plain Text Message Test")
print("=" * 80)

# Load credentials
cred_dir = Path.home() / ".openclaw" / "credentials" / "awiki-agent-id-message"

# Use bearertest (fresh JWT) and pythonagent
sender_cred_path = cred_dir / "bearertest.json"
receiver_cred_path = cred_dir / "pythonagent.json"

sender_cred = json.loads(sender_cred_path.read_text())
receiver_cred = json.loads(receiver_cred_path.read_text())

sender_did = sender_cred["did"]
sender_jwt = sender_cred["jwt_token"]
receiver_did = receiver_cred["did"]

print(f"\nSender: {sender_did}")
print(f"Receiver: {receiver_did}")
print(f"Sender JWT: {sender_jwt[:50]}...")

async def send_message(sender_jwt, sender_did, receiver_did, content):
    """Send a plain text message."""
    request_body = {
        "jsonrpc": "2.0",
        "method": "send",
        "params": {
            "sender_did": sender_did,
            "receiver_did": receiver_did,
            "content": content,
            "type": "text"
        },
        "id": 1
    }
    
    async with httpx.AsyncClient(timeout=30.0) as client:
        response = await client.post(
            "https://awiki.ai/message/rpc",
            json=request_body,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {sender_jwt}"
            }
        )
        
        return response.json()

async def get_inbox(jwt, user_did, limit=10):
    """Get inbox messages."""
    request_body = {
        "jsonrpc": "2.0",
        "method": "get_inbox",
        "params": {
            "user_did": user_did,
            "limit": limit
        },
        "id": 1
    }
    
    async with httpx.AsyncClient(timeout=30.0) as client:
        response = await client.post(
            "https://awiki.ai/message/rpc",
            json=request_body,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {jwt}"
            }
        )
        
        return response.json()

async def main():
    # Test 1: Send message from bearertest to pythonagent
    print("\n" + "-" * 80)
    print("Test 1: Send message (bearertest -> pythonagent)")
    print("-" * 80)
    
    message_content = f"Hello from bearertest! This is a plain text message test."
    print(f"Message: {message_content}")
    
    send_result = await send_message(sender_jwt, sender_did, receiver_did, message_content)
    
    if send_result.get("result"):
        print("Send: SUCCESS")
        print(f"Message ID: {send_result['result'].get('id', 'N/A')}")
        print(f"Server Seq: {send_result['result'].get('server_seq', 'N/A')}")
    else:
        print("Send: FAILED")
        print(f"Error: {send_result.get('error', {}).get('message', 'Unknown error')}")
        return
    
    # Wait for message to be delivered
    await asyncio.sleep(1)
    
    # Test 2: Get inbox for receiver
    print("\n" + "-" * 80)
    print("Test 2: Get inbox (pythonagent)")
    print("-" * 80)
    
    inbox_result = await get_inbox(receiver_cred["jwt_token"], receiver_did)
    
    if inbox_result.get("result"):
        print("Get Inbox: SUCCESS")
        messages = inbox_result["result"].get("messages", [])
        print(f"Total messages: {len(messages)}")
        
        if messages:
            print("\nLatest messages:")
            for msg in messages[:5]:
                print(f"  [{msg.get('type', 'N/A')}] From: {msg.get('sender_did', 'N/A')[:30]}...")
                print(f"    Content: {msg.get('content', 'N/A')[:50]}...")
                print(f"    Server Seq: {msg.get('server_seq', 'N/A')}")
                
                # Check if our message is in the inbox
                if msg.get('content') == message_content:
                    print("\n  ✓ Our test message found in inbox!")
        else:
            print("No messages in inbox")
    else:
        print("Get Inbox: FAILED")
        print(f"Error: {inbox_result.get('error', {}).get('message', 'Unknown error')}")
    
    # Test 3: Send message back (pythonagent -> bearertest)
    print("\n" + "-" * 80)
    print("Test 3: Send reply (pythonagent -> bearertest)")
    print("-" * 80)
    
    reply_content = f"Reply from pythonagent! Received your message."
    print(f"Message: {reply_content}")
    
    reply_result = await send_message(receiver_cred["jwt_token"], receiver_did, sender_did, reply_content)
    
    if reply_result.get("result"):
        print("Send Reply: SUCCESS")
        print(f"Message ID: {reply_result['result'].get('id', 'N/A')}")
    else:
        print("Send Reply: FAILED")
        print(f"Error: {reply_result.get('error', {}).get('message', 'Unknown error')}")
    
    # Summary
    print("\n" + "=" * 80)
    print("SUMMARY")
    print("=" * 80)
    print("Plain text message test completed.")
    print("Check the output above for details.")

asyncio.run(main())
